### ISSUE TITLE
Distinguish between compiler failure and user code error in compiler return code

### ISSUE THREAD

**Issue by bburdette** [elm/compiler#1981]
**Quick Summary:**

I'd like to be able to distinguish between a compiler failure and an error in my elm code.  Then I could write a bash script to automatically delete elm-stuff and retry.  

For example, both of these scenarios return the same error, '1':  

```
-- PARSE ERROR -------------------------------------------------- src/Dialog.elm

Something went wrong while parsing your code.

31| dmMap : DMsg a -> (a -> b) -> DMsg b
    ^
I was expecting to see more indentation? I was not done with that last thing
yet.
```
The return code from the executable here is '1'.

```
elm: not enough bytes
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary.hs:212:21 in binary-0.8.6.0:Data.Binary
```
The return code here is also '1'.  If it was '2', or '127' I could remove the elm-stuff and try again.

**Comment by Razzeee** [comment]
Feels very hacky, removing `elm-stuff` does not gurantee your problem to be gone.

Anyway, did you try to run `elm make` with `--report=json`? I think that has at least two types of errors `compile-errors` and `error`.
https://github.com/elm/compiler/blob/822401199c103a1567f2d208a9e8014740e4732e/builder/src/Reporting/Exit/Help.hs#L94

**Comment by bburdette** [comment]
Yep its a hack, but in my case its pretty consistent - if there was a code error last time I compiled, its 90% certain I'm going to need to delete elm-stuff, in one particular project anyway.  

For now I can tee the output into a file and search that file for the not-enough-bytes message.  report=json would be better for that, but then I wouldn't have the human readable output.

**Comment by Razzeee** [comment]
What I'm trying to say is, I don't believe that the compiler should cater to a hack. Better would be to fix the underlying issue.

The error you mentioned here should be fixed in 0.19.1

**Comment by bburdette** [comment]
Fair enough!  I have a solution that works for me for now.

**Comment by bburdette** [comment]
One point to note here is the compiler error doesn't produce valid json for me, at least on 0.19.0.  Its just the regular human readable error.

**Comment by evancz** [comment]
Is the "not enough bytes" issue resolved with 0.19.1?

(As an aside, this appears to happen for 0.19.0 users who have VSCode or some editor where it runs multiple `elm` calls simultaneously for a single project. Even if this is fixed by 0.19.1 the authors of the relevant plugins or build tools should avoid doing this!)

**Comment by bburdette** [comment]
I haven't upgraded yet!  I'll check it soon.

**Comment by bburdette** [comment]
Ok I upgraded and its looking good!  No compile errors so far, after a bunch of edit-build cycles.

Also the compile is WAY faster than before - my problem project built pretty slowly under 0.19.0, but the incremental build on 0.19.1 is super quick.  I have some "type classes" with type parameters that are used all over the place and I blame those (without hard evidence) for the previous slow compiling.  That seems to be a non-issue now.  Under 0.19.1, the initial compile did take about as long as 0.19.0 - 15 or 20 seconds - but the incremental builds are almost instantaneous.  Sweet!
